rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,634 @@
+  // GetRotationState returns whether or not a key rotation is in progress.
+  rpc GetRotationState(GetRotationStateRequest) returns (GetRotationStateResponse);
+  // CompleteRotation moves rotated keys out of the active set.
+  rpc CompleteRotation(CompleteRotationRequest) returns (CompleteRotationResponse);
+  // UploadEncryptedRecording is used to upload encrypted .tar files
+  // containing session recording events into long term storage.
+  rpc UploadEncryptedRecording(stream UploadEncryptedRecordingRequest) returns (UploadEncryptedRecordingResponse);
+}
+
+// RotateKeySetRequest
+message RotateKeySetRequest {}
+
+// RotateKeySetResponse
+message RotateKeySetResponse {}
+
+// GetRotationStateRequest
+message GetRotationStateRequest {}
+
+// AgeEncryptionKeyWithState reports the KeyState for a given public key.
+message AgeEncryptionKeyWithState {
+  // PublicKey is the Bech32 encoded age X25519 public key associated with the KeyState.
+  bytes public_key = 1;
+  // KeyState is the state PublicKey is currently in.
+  teleport.recordingencryption.v1.KeyState key_state = 2;
+}
+
+// GetRotationStateResponse contains the current KeyState of all WrappedKeys
+// in the EncryptedSessionRecordingConfig.
+message GetRotationStateResponse {
+  // Keys contains the list of currently active public keys along with their status.
+  repeated AgeEncryptionKeyKeyWithState keys = 1;
+}
+
+// CompleteRotationRequest
+message CompleteRotationRequest {}
+
+// CompleteRotationResponse
+message CompleteRotationResponse {}
+
+// UploadEncryptedRecordingRequest is an individual part of an encrypted session
+// recording .tar file.
+message UploadEncryptedRecordingRequest {
+  // SessionID the recording relates to.
+  string session_id = 1;
+  // PartIndex is the ordered index applied to the chunk.
+  int64 part_index = 2;
+  // Part is the encrypted part of session recording data being uploaded.
+  bytes part = 3;
+}
+
+// UploadEncryptedRecordingResponse
+message UploadEncryptedRecordingResponse {}
+```
+
+### Session Recording Modes
+
+There are four session recording modes that describe where the recordings are
+captured and how they're shipped to long term storage.
+
+- `proxy-sync`
+- `proxy`
+- `node-sync`
+- `node`
+  Where the recordings are collected is largely unimportant to the encryption
+  strategy, but whether or not they are handled async or sync has different
+  requirements.
+
+In sync modes the session recording data is written immediately to the auth
+service without intermediate disk writes. The auth service then handles
+batching and multipart upload of events to long term storage. In order to
+ensure resilience to partial uploads of parts, each batch will need to be
+encrypted individually on the auth service. This will likely mean wrapping the
+`gzipWriter` [used by slices](https://github.com/gravitational/teleport/blob/master/lib/events/stream.go#L669)
+with an `age` encryption writer.
```

Comment:
Parts must be padded up to 5MiB (unless they're the last part) to support S3 multipart uploads, are we going to do that outside of the age envelope or can we do padding inside and get to exactly 5MiB on the outside? i.e. if we add one byte of plaintext, will we have one more byte of ciphertext?

Reply:
The padding will be outside of the age envelope, so the file contents are:
```
24-byte header | encrypted, gzipped proto messages | padding up to 5MiB if required
```

Reply:
Yeah definitely. I wasn't originally planning on changing the header format since we can inspect that the data segment starts with the `age` header, but after implementing that yesterday I think it might be better to add an extra byte for a bitflag that stores whether or not the data is encrypted. It looks like the header is intentionally 8 byte aligned, so I'd probably pad another 7 bytes of "reserved" space. I'll add this to the RFD